### PR TITLE
Conditional Branch Tests

### PR DESCRIPTION
**Pull request by vext01**
c27f1dd is the main work here.

The way I've generated the tests is more principled this time (systematically chooses boundary values for all types for each operator). After this is merged I'd like to go back and refactor the binop tests to do the same.

I also realised that we also don't have to provide our own expected outcomes, as they can be computed.

This fixes a bug that @ptersilie mentioned too. I've been trying to drill this into my head:
image

9b6657c is just cosmetic: moves some existing tests into a new file.

### CODE REVIEW

File: tests/src/codegen/cond_brs.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                let th = start_tracing(TracingKind::HardwareTracing);
+                #[cfg(tracermode = "sw")]
+                let th = start_tracing(TracingKind::SoftwareTracing);
+                interp_step(&mut ctx);
+                let sir_trace = th.stop_tracing().unwrap();
+                let ct = compile_trace(sir_trace).unwrap();
+
+                let mut args = CmpCtx::new($arg1, $arg2, 255);
+                assert!(unsafe { ct.execute(&mut args).is_null() });
+                println!("{} {} {} for {}: {}", stringify!($arg1), stringify!($op), stringify!($arg2), stringify!($ty), $arg1 $op $arg2);
```
Comment (ptersilie) [inline tests/src/codegen/cond_brs.rs:50]: Was this print left here on purpose?

File: tests/src/codegen/cond_brs.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+                impl CmpCtx {
+                    fn new(arg1: $type, arg2: $type, res: u8) -> Self {
+                        Self { arg1, arg2, res }
+                    }
+                }
+
+                #[interp_step]
+                fn interp_step(ctx: &mut CmpCtx) -> bool {
+                    if ctx.arg1 $op ctx.arg2 {
```
[the remark this answers is not in the document]
Comment (vext01) [reply tests/src/codegen/cond_brs.rs:31]: I considered this, but I didn't want the compiler to remove the conditional branch.

It could change:
```
if ctx.arg1 $op ctx.arg2 {
  ctx.res = 1
} else {
  ctx.res = 0
}
```

into:

```
ctx.res = ctx.arg1 $op ctx.arg2;
```

File: tests/src/codegen/cond_brs.rs (hunk 1 above)
Comment (vext01) [reply tests/src/codegen/cond_brs.rs:50]: hah. no, my bad.

Comment (vext01) [reply tests/src/codegen/cond_brs.rs]: Fixed in 21c93379e

File: tests/src/codegen/cond_brs.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ptersilie) [reply tests/src/codegen/cond_brs.rs:31]: Isn't that the same I suggested above?

[the remark this answers is not in the document]
Comment (vext01) [reply tests/src/codegen/cond_brs.rs:31]: Yes, but the whole point of these tests are to check that conditional branching is working. So we need to test code that will have a conditional branch :)

[the remark this answers is not in the document]
Comment (ptersilie) [reply tests/src/codegen/cond_brs.rs:31]: Got you!

**Comment by ptersilie** [comment]
Please squash.

**Comment by vext01** [comment]
splat.
